### DISCUSSION TITLE
LESSON :7  testFundUpatesFundedDataStructure ()      ERROR:   [FAIL. Reason: cannot `prank` for a broadcasted transaction; pass the desired `tx.origin` into the `broadcast` cheatcode call]

### DISCUSSION THREAD

**Discussion by Basha-dude** [Cyfrin/foundry-full-course-cu#1576] [General]
FOR THIS CODE ```// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../src/FundMe.sol";
import {DeployFundMe} from "../script/DeployFundMe.s.sol";

contract FundMeTest is Test {
    FundMe fundme;

    address USER = makeAddr("user");
      
    uint256 constant SEND_VALUE = 0.1 ether;
    uint256 constant STARTING_BALANCE = 10 ether;

    function setUp() external {
        DeployFundMe deployFundMe = new DeployFundMe();
        fundme = deployFundMe.run();
            vm.deal(USER, STARTING_BALANCE);  
              }

    function testOwner() public view {
        console.log("ADDRESS of fundme ", address(fundme));
        console.log("msg.sender ", msg.sender);
        assertEq(fundme.i_owner(), msg.sender);
    }

    function testMinimumUsdIsFive() public view {
        assertEq(fundme.MINIMUM_USD(), 5e18);
        console.log("msg.sender of testMinimumUsdIsFive ", msg.sender);
    }

    function testFund() public payable {
        // fundme.fund{value: 10000000000000000}();
        // address[]  memory funder = fundme.funders;
        // assertEq(funder[0],msg.sender);
    }

    function testPriceFeedversionisAccurate() public view {
        uint256 version = fundme.getVersion();
        assertEq(version, 4);
    }

    function testFundFailsWithoutEth() public {
        vm.expectRevert();
        fundme.fund();
    }

     function testFundUpatesFundedDataStructure()  public {
        vm.prank(USER);
        fundme.fund{value: SEND_VALUE}();
        uint256 amountFunded = fundme.getAddressToAmountFunded(USER);
        assertEq(amountFunded,SEND_VALUE);
     }
}
``` 

ERRO:

GIVING THIS ERROR  ```~/f23/foundry-fund-me$ forge test --mt testFundUpatesFundedDataStructure
[⠢] Compiling...
No files changed, compilation skipped

Ran 1 test for test/FundMe.t.sol:FundMeTest
[FAIL. Reason: cannot `prank` for a broadcasted transaction; pass the desired `tx.origin` into the `broadcast` cheatcode call] testFundUpatesFundedDataStructure() (gas: 5163)
Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 18.70ms (825.80µs CPU time)

Ran 1 test suite in 37.30ms (18.70ms CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/FundMe.t.sol:FundMeTest
[FAIL. Reason: cannot `prank` for a broadcasted transaction; pass the desired `tx.origin` into the `broadcast` cheatcode call] testFundUpatesFundedDataStructure() (gas: 5163)

Encountered a total of 1 failing tests, 0 tests succeeded```

**Comment by EngrPips** [comment]
Hello @Basha-dude , please push your codebase to github and share the github link with me so that i can clone, set it up locally on my machine and try debug it. It hard for me to figure out what is wrong merely by looking at the code you share

**Reply by Basha-dude** [reply]
thanks for helping and this is the link 
https://github.com/Basha-dude/foundry-7-vm.prank-error

**Reply by EngrPips** [reply]
Awesome, thank you so much. i would clone this look into it and get back to you with whatever it is that i find.

**Reply by EngrPips** [reply]
Hello @Basha-dude , so i found out that the issue was with your `DeployFundMe.s.sol` script, when you wanted to deploy a new `FundMe` contract in the script you started a broadcast with `vm.startBroadcast()` but you didn't stop the broadcast with `vm.stopBroadcast()` after you finish deploying, hence the broadcast was still active and then in your test you tried to use `vm.startPrank(USER)` while a broadcast was still active from  `DeployFundMe.s.sol`, broadcast and prank are not compatible hence  your test fails. 

<img width="1440" alt="Screenshot 2024-04-05 at 9 08 05 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/f57f32be-07b2-4177-b4fd-bceddf6f0c67">

as shown in the image above, in your `DeployFundMe.s.sol` after the `FundMe fundMe = new FundMe(pricefeed);` line add `vm.stopBroadcast()` and rerun your test and it should pass.

I am looking forward to hear back from you.

**Reply by Basha-dude** [reply]
yes, you are correct, there is no error, thank you very much for helping me out

**Reply by EngrPips** [reply]
Anytime, i am glad you are able to fix it. keep crushing it buddy!

**Reply by Basha-dude** [reply]
"Hey buddy, I got into a problem. can you help me

**Reply by Basha-dude** [reply]
and this is the link [https://github.com/Basha-dude/foundry-security-course-errorGiving-different](https://github.com/Basha-dude/foundry-security-course-errorGiving-different).

**Reply by EngrPips** [reply]
I replied you @Basha-dude.
